File: packages/sync/src/medicineSearch.ts
/** Every field the server may report; the local matcher returns a subset. */
export type MedicineMatchField = 'barcode' | 'sku' | 'name' | 'genericName' | 'alias' | 'strength' | 'dosageForm';
export type MedicineMatchQuality = 'exact' | 'partial' | 'fuzzy' | 'supporting';

export type MedicineSearchValue = {
  name: string;
  genericName?: string | null;
  strength?: string | null;
  dosageForm?: string | null;
  manufacturerId?: string | null;
  manufacturer?: string | null;
};

export type MedicineTextMatch = {
  matchedField: MedicineMatchField;
  matchQuality: MedicineMatchQuality;
  matchedText: string;
  matchScore: number;
  /** Lower is safer/more relevant. Identifier ranks 0 and 1 are owned by shelf.ts. */
  rank: number;
};

export type RankedMedicine<T extends MedicineSearchValue> = MedicineTextMatch & { item: T };

export type MedicineDosageGroup<T extends MedicineSearchValue> = {
  key: string;
  label: string;
  bestRank: number;
  bestScore: number;
  items: readonly RankedMedicine<T>[];
};

export type MedicineManufacturerGroup<T extends MedicineSearchValue> = {
  key: string;
  label: string;
  bestRank: number;
  bestScore: number;
  count: number;
  dosageGroups: readonly MedicineDosageGroup<T>[];
};

const UNIT_PATTERN = /\b\d+(?:\.\d+)?\s*(?:mcg|mg|gm|g|ml|l|iu|units?|meq|mmol|%)\b/giu;

/** Stable normalization shared by all offline counter shells. */
export function normalizeMedicineText(value: string): string {
  return value
    .normalize('NFKC')
    .toLocaleLowerCase()
    .replace(/[‐‑‒–—−]/g, '-')
    .replace(/[^\p{L}\p{N}_+/%.-]+/gu, ' ')
    .replace(/\s+/g, ' ')
    .trim()
    .replace(/(\d)\s+(mcg|mg|gm|g|ml|l|iu|unit|units|meq|mmol|%)(?=\b|\/)/giu, '$1$2');
}

function editBudget(length: number): number {
  if (length < 3) return 0;
  if (length <= 5) return 1;
  if (length <= 9) return 2;
  return 3;
}

/** Optimal-string-alignment Damerau-Levenshtein; adjacent transposition costs one. */
export function medicineEditDistance(leftRaw: string, rightRaw: string): number {
  const left = Array.from(normalizeMedicineText(leftRaw));
  const right = Array.from(normalizeMedicineText(rightRaw));
  const rows = Array.from({ length: left.length + 1 }, () => Array<number>(right.length + 1).fill(0));
  for (let i = 0; i <= left.length; i += 1) rows[i]![0] = i;
  for (let j = 0; j <= right.length; j += 1) rows[0]![j] = j;
  for (let i = 1; i <= left.length; i += 1) {
    for (let j = 1; j <= right.length; j += 1) {
      const substitution = left[i - 1] === right[j - 1] ? 0 : 1;
      let distance = Math.min(
        rows[i - 1]![j]! + 1,
        rows[i]![j - 1]! + 1,
        rows[i - 1]![j - 1]! + substitution,
      );
      if (
        i > 1 && j > 1 &&
        left[i - 1] === right[j - 2] &&
        left[i - 2] === right[j - 1]
      ) {
        distance = Math.min(distance, rows[i - 2]![j - 2]! + 1);
      }
      rows[i]![j] = distance;
    }
  }
  return rows[left.length]![right.length]!;
}

function fuzzyScore(query: string, field: string): number | null {
  if (query.length < 3) return null;
  const words = field.split(' ').filter(Boolean);
  const queryWords = query.split(' ').filter(Boolean);
  const width = Math.max(1, queryWords.length);
  const candidates = new Set<string>([field]);
  for (let index = 0; index < words.length; index += 1) {
    candidates.add(words.slice(index, index + width).join(' '));
  }
  let best: number | null = null;
  for (const candidate of candidates) {
    if (candidate === '') continue;
    const distance = medicineEditDistance(query, candidate);
    const score = 1 - distance / Math.max(Array.from(query).length, Array.from(candidate).length);
    if (distance <= editBudget(Array.from(query).length) && score >= 0.70) {
      best = best === null ? score : Math.max(best, score);
    }
  }
  return best;
}

function removePhrase(value: string, phrase: string): string {
  if (phrase === '') return value;
  return value
    .replace(new RegExp(`(^|\\s)${escapeRegex(phrase)}(?=\\s|$)`, 'gu'), ' ')
    .replace(/\s+/g, ' ')
    .trim();
}

function escapeRegex(value: string): string {
  return value.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

function queryForCandidate(query: string, item: MedicineSearchValue): { text: string; supports: MedicineTextMatch[]; rejected: boolean } {
  let text = query;
  const supports: MedicineTextMatch[] = [];
  const dosage = normalizeMedicineText(item.dosageForm ?? '');
  if (dosage !== '' && (` ${text} `).includes(` ${dosage} `)) {
    text = removePhrase(text, dosage);
    supports.push({ matchedField: 'dosageForm', matchQuality: 'supporting', matchedText: item.dosageForm ?? '', matchScore: 1, rank: 10 });
  }

  const requestedStrengths = Array.from(query.matchAll(UNIT_PATTERN), (match) => normalizeMedicineText(match[0]));
  const strength = normalizeMedicineText(item.strength ?? '');
  if (requestedStrengths.length > 0) {
    if (strength === '' || requestedStrengths.some((term) => !strength.includes(term))) {
      return { text, supports, rejected: true };
    }
    for (const term of requestedStrengths) text = removePhrase(text, term);
    supports.push({ matchedField: 'strength', matchQuality: 'supporting', matchedText: item.strength ?? '', matchScore: 1, rank: 10 });
  }
  if (strength !== '') {
    // A bare "500" ("napa 500 tablet") is a strength the cashier left the unit off
    // of. It matches only the digits written on the row: 500 does not stand in for
    // 650, and a row with no strength at all keeps the number for its name.
    const strengthNumbers = new Set(Array.from(strength.matchAll(/\d+(?:\.\d+)?/g), (match) => match[0]));
    for (const token of text.split(' ')) {
      if (!/^\d+(?:\.\d+)?$/.test(token)) continue;
      if (!strengthNumbers.has(token)) return { text, supports, rejected: true };
      text = removePhrase(text, token);
      supports.push({ matchedField: 'strength', matchQuality: 'supporting', matchedText: item.strength ?? '', matchScore: 1, rank: 10 });
    }
  }
  return { text, supports, rejected: false };
}

/** Rank one medicine by brand/generic text after applying exact strength/form support terms. */
export function matchMedicineText(item: MedicineSearchValue, rawQuery: string): MedicineTextMatch | null {
  const query = normalizeMedicineText(rawQuery);
  if (query === '') return null;
  const name = normalizeMedicineText(item.name);
  const generic = normalizeMedicineText(item.genericName ?? '');

  // A custom product often embeds its strength in its local name, so full-field
  // checks happen before candidate support terms are removed.
  if (name === query) return { matchedField: 'name', matchQuality: 'exact', matchedText: item.name, matchScore: 1, rank: 2 };
  if (generic !== '' && generic === query) return { matchedField: 'genericName', matchQuality: 'exact', matchedText: item.genericName ?? '', matchScore: 1, rank: 3 };

  const parsed = queryForCandidate(query, item);
  if (parsed.rejected) return null;
  const text = parsed.text;
  if (text === '') return parsed.supports[0] ?? null;

  if (name === text) return { matchedField: 'name', matchQuality: 'exact', matchedText: item.name, matchScore: 1, rank: 2 };
  if (generic !== '' && generic === text) return { matchedField: 'genericName', matchQuality: 'exact', matchedText: item.genericName ?? '', matchScore: 1, rank: 3 };
  if (name.includes(text)) return { matchedField: 'name', matchQuality: 'partial', matchedText: item.name, matchScore: text.length / name.length, rank: 5 };
  if (generic !== '' && generic.includes(text)) return { matchedField: 'genericName', matchQuality: 'partial', matchedText: item.genericName ?? '', matchScore: text.length / generic.length, rank: 6 };

  const nameScore = fuzzyScore(text, name);
  const genericScore = generic === '' ? null : fuzzyScore(text, generic);
  if (nameScore !== null) return { matchedField: 'name', matchQuality: 'fuzzy', matchedText: item.name, matchScore: nameScore, rank: 8 };
  if (genericScore !== null) return { matchedField: 'genericName', matchQuality: 'fuzzy', matchedText: item.genericName ?? '', matchScore: genericScore, rank: 9 };
  return null;
}

function compareRanked<T extends MedicineSearchValue>(left: RankedMedicine<T>, right: RankedMedicine<T>): number {
  return left.rank - right.rank || right.matchScore - left.matchScore || left.item.name.localeCompare(right.item.name);
}

/** Manufacturer -> dosage hierarchy; each group inherits its best child's rank. */
export function groupMedicineMatches<T extends MedicineSearchValue>(matches: readonly RankedMedicine<T>[]): readonly MedicineManufacturerGroup<T>[] {
  const manufacturers = new Map<string, { label: string; matches: RankedMedicine<T>[] }>();
  for (const match of matches) {
    const manufacturer = match.item.manufacturer?.trim() || 'Custom / manufacturer not set';
    const key = match.item.manufacturerId ?? `label:${normalizeMedicineText(manufacturer)}`;
    const bucket = manufacturers.get(key) ?? { label: manufacturer, matches: [] };
    bucket.matches.push(match);
    manufacturers.set(key, bucket);
  }

  const groups: MedicineManufacturerGroup<T>[] = [];
  for (const [key, manufacturer] of manufacturers) {
    const dosages = new Map<string, { label: string; matches: RankedMedicine<T>[] }>();
    for (const match of manufacturer.matches) {
      const dosage = match.item.dosageForm?.trim() || 'Dosage form not set';
      const dosageKey = `label:${normalizeMedicineText(dosage)}`;
      const bucket = dosages.get(dosageKey) ?? { label: dosage, matches: [] };
      bucket.matches.push(match);
      dosages.set(dosageKey, bucket);
    }
    const dosageGroups = Array.from(dosages, ([dosageKey, dosage]) => {
      const items = dosage.matches.sort(compareRanked);
      return { key: dosageKey, label: dosage.label, bestRank: items[0]!.rank, bestScore: items[0]!.matchScore, items };
    }).sort((a, b) => a.bestRank - b.bestRank || b.bestScore - a.bestScore || a.label.localeCompare(b.label));
    groups.push({
      key,
      label: manufacturer.label,
      bestRank: dosageGroups[0]!.bestRank,
      bestScore: dosageGroups[0]!.bestScore,
      count: manufacturer.matches.length,
      dosageGroups,
    });
  }
  return groups.sort((a, b) => a.bestRank - b.bestRank || b.bestScore - a.bestScore || a.label.localeCompare(b.label));
}

/**
 * The rank a server search row carries, derived from the metadata
 * `GET /products/search` reports. The same tier order the Python ranker uses;
 * shared here so the catalogue screen groups API rows exactly as the POS shells
 * group local shelf matches.
 */
const MEDICINE_MATCH_RANKS: Record<string, number> = {
  'barcode:exact': 0,
  'sku:exact': 1,
  'name:exact': 2,
  'genericName:exact': 3,
  'alias:exact': 4,
  'name:partial': 5,
  'genericName:partial': 6,
  'alias:partial': 7,
  'name:fuzzy': 8,
  'genericName:fuzzy': 9,
  'strength:supporting': 10,
  'dosageForm:supporting': 10,
};

export function medicineMatchRank(field: MedicineMatchField, quality: MedicineMatchQuality): number {
  return MEDICINE_MATCH_RANKS[`${field}:${quality}`] ?? 10;
}

/** What a row says about why it is on the screen, in a cashier's words. */
export function describeMedicineMatch(match: { matchedField: MedicineMatchField; matchQuality: MedicineMatchQuality }): string {
  const key = `${match.matchedField}:${match.matchQuality}`;
  if (key === 'barcode:exact') return 'Exact barcode';
  if (key === 'sku:exact') return 'Exact SKU';
  if (key === 'name:exact') return 'Exact brand match';
  if (key === 'genericName:exact') return 'Exact generic match';
  if (key === 'alias:exact') return 'Exact alias match';
  if (key === 'name:partial') return 'Brand name match';
  if (key === 'genericName:partial') return 'Generic name match';
  if (key === 'alias:partial') return 'Alias match';
  if (key === 'name:fuzzy') return 'Closest brand match';
  if (key === 'genericName:fuzzy') return 'Closest generic match';
  if (key === 'strength:supporting') return 'Matched by strength';
  return 'Matched by dosage form';
}

/**
 * True when nothing on the list is better than a typo guess -- the screen owes
 * the cashier the "no exact match" banner rather than a list that reads as
 * confirmed results.
 */
export function medicineMatchesAreFuzzy(matches: readonly { matchQuality: MedicineMatchQuality }[]): boolean {
  return matches.length > 0 && matches.every((match) => match.matchQuality === 'fuzzy');
}

export type HighlightSpan = { text: string; hit: boolean };

/**
 * Split `text` around literal occurrences of the query's tokens, so a screen can
 * mark exactly what was typed. Matching is plain case-insensitive equality of
 * substrings -- no fuzzy hits are highlighted, because a mark under a word the
 * cashier did not type is a lie about why the row appeared.
 */
export function highlightMedicineSpans(text: string, query: string): readonly HighlightSpan[] {
  const tokens = Array.from(new Set(query.split(/\s+/).map((token) => token.trim()).filter((token) => token !== '')))
    .sort((left, right) => right.length - left.length);
  if (tokens.length === 0 || text === '') return [{ text, hit: false }];

  const haystack = text.toLocaleLowerCase();
  const ranges: Array<[number, number]> = [];
  for (const token of tokens) {
    const needle = token.toLocaleLowerCase();
    let from = haystack.indexOf(needle);
    while (from !== -1) {
      ranges.push([from, from + needle.length]);
      from = haystack.indexOf(needle, from + needle.length);
    }
  }
  if (ranges.length === 0) return [{ text, hit: false }];

  ranges.sort((left, right) => left[0] - right[0] || left[1] - right[1]);
  const merged: Array<[number, number]> = [ranges[0]!];
  for (const range of ranges.slice(1)) {
    const last = merged[merged.length - 1]!;
    if (range[0] <= last[1]) last[1] = Math.max(last[1], range[1]);
    else merged.push(range);
  }

  const spans: HighlightSpan[] = [];
  let cursor = 0;
  for (const [start, end] of merged) {
    if (cursor < start) spans.push({ text: text.slice(cursor, start), hit: false });
    spans.push({ text: text.slice(start, end), hit: true });
    cursor = end;
  }
  if (cursor < text.length) spans.push({ text: text.slice(cursor), hit: false });
  return spans;
}

File: backend/app/services/medicine_search.py
from __future__ import annotations

import re
import unicodedata
from dataclasses import dataclass
from typing import Literal

MatchField = Literal["barcode", "sku", "name", "genericName", "alias", "strength", "dosageForm"]
MatchQuality = Literal["exact", "partial", "fuzzy", "supporting"]

_UNITS = r"(?:mcg|mg|gm|g|ml|l|iu|units?|meq|mmol|%)"
_UNIT_RE = re.compile(rf"\b\d+(?:\.\d+)?\s*{_UNITS}\b", re.IGNORECASE)
_SEPARATORS_RE = re.compile(r"[^\w+/%.-]+", re.UNICODE)
_SPACES_RE = re.compile(r"\s+")
_UNIT_SPACE_RE = re.compile(rf"(\d)\s+({_UNITS[3:-1]})(?=\b|/)", re.IGNORECASE)
#: The digits inside a strength value, so "500" in a query can be checked
#: against "500 mg" without the unit having to be typed.
_STRENGTH_NUMBER_RE = re.compile(r"\d+(?:\.\d+)?")


@dataclass(frozen=True, slots=True)
class MedicineMatch:
    field: MatchField
    quality: MatchQuality
    text: str
    score: float
    rank: int


def normalize_medicine_text(value: str) -> str:
    normalized = unicodedata.normalize("NFKC", value).casefold()
    normalized = normalized.translate(str.maketrans({"‐": "-", "‑": "-", "‒": "-", "–": "-", "—": "-", "−": "-"}))
    normalized = _SPACES_RE.sub(" ", _SEPARATORS_RE.sub(" ", normalized)).strip()
    return _UNIT_SPACE_RE.sub(r"\1\2", normalized)


def medicine_edit_distance(left_raw: str, right_raw: str) -> int:
    left = list(normalize_medicine_text(left_raw))
    right = list(normalize_medicine_text(right_raw))
    matrix = [[0] * (len(right) + 1) for _ in range(len(left) + 1)]
    for index in range(len(left) + 1):
        matrix[index][0] = index
    for index in range(len(right) + 1):
        matrix[0][index] = index
    for row in range(1, len(left) + 1):
        for column in range(1, len(right) + 1):
            substitution = 0 if left[row - 1] == right[column - 1] else 1
            distance = min(
                matrix[row - 1][column] + 1,
                matrix[row][column - 1] + 1,
                matrix[row - 1][column - 1] + substitution,
            )
            if (
                row > 1
                and column > 1
                and left[row - 1] == right[column - 2]
                and left[row - 2] == right[column - 1]
            ):
                distance = min(distance, matrix[row - 2][column - 2] + 1)
            matrix[row][column] = distance
    return matrix[-1][-1]


def _edit_budget(length: int) -> int:
    if length < 3:
        return 0
    if length <= 5:
        return 1
    if length <= 9:
        return 2
    return 3


def _fuzzy_score(query: str, field: str) -> float | None:
    if len(query) < 3:
        return None
    words = field.split()
    width = max(1, len(query.split()))
    candidates = {field, *(" ".join(words[index : index + width]) for index in range(len(words)))}
    accepted: list[float] = []
    for candidate in candidates:
        if not candidate:
            continue
        distance = medicine_edit_distance(query, candidate)
        score = 1 - distance / max(len(query), len(candidate))
        if distance <= _edit_budget(len(query)) and score >= 0.70:
            accepted.append(score)
    return max(accepted) if accepted else None


def _remove_phrase(value: str, phrase: str) -> str:
    if not phrase:
        return value
    removed = re.sub(rf"(^|\s){re.escape(phrase)}(?=\s|$)", " ", value)
    return _SPACES_RE.sub(" ", removed).strip()


def _bare_number_terms(value: str) -> list[str]:
    """Standalone number tokens: "500" in "napa 500", but not in "b12"."""
    return [token for token in value.split(" ") if _BARE_NUMBER_TOKEN_RE.fullmatch(token)]


_BARE_NUMBER_TOKEN_RE = re.compile(r"\d+(?:\.\d+)?")


def search_core(raw_query: str, dosage_forms: list[tuple[object, str]]) -> tuple[str, object | None, tuple[str, ...]]:
    """Split globally recognizable dosage/strength support terms from medicine text."""
    query = normalize_medicine_text(raw_query)
    matching_forms = [
        (form_id, normalize_medicine_text(name))
        for form_id, name in dosage_forms
        if normalize_medicine_text(name) and f" {normalize_medicine_text(name)} " in f" {query} "
    ]
    matching_forms.sort(key=lambda row: len(row[1]), reverse=True)
    dosage_id: object | None = None
    if matching_forms:
        dosage_id, dosage_text = matching_forms[0]
        query = _remove_phrase(query, dosage_text)
    strengths = tuple(normalize_medicine_text(match.group(0)) for match in _UNIT_RE.finditer(query))
    for strength in strengths:
        query = _remove_phrase(query, strength)
    return query, dosage_id, strengths


def match_medicine_text(
    *,
    name: str,
    generic_name: str | None,
    strength: str | None,
    dosage_form: str | None,
    raw_query: str,
) -> MedicineMatch | None:
    query = normalize_medicine_text(raw_query)
    if not query:
        return None
    normalized_name = normalize_medicine_text(name)
    normalized_generic = normalize_medicine_text(generic_name or "")
    if normalized_name == query:
        return MedicineMatch("name", "exact", name, 1, 2)
    if normalized_generic and normalized_generic == query:
        return MedicineMatch("genericName", "exact", generic_name or "", 1, 3)

    remaining = query
    support: list[MedicineMatch] = []
    normalized_dosage = normalize_medicine_text(dosage_form or "")
    if normalized_dosage and f" {normalized_dosage} " in f" {remaining} ":
        remaining = _remove_phrase(remaining, normalized_dosage)
        support.append(MedicineMatch("dosageForm", "supporting", dosage_form or "", 1, 10))
    requested_strengths = tuple(normalize_medicine_text(match.group(0)) for match in _UNIT_RE.finditer(query))
    normalized_strength = normalize_medicine_text(strength or "")
    if requested_strengths:
        if not normalized_strength or any(term not in normalized_strength for term in requested_strengths):
            return None
        for term in requested_strengths:
            remaining = _remove_phrase(remaining, term)
        support.append(MedicineMatch("strength", "supporting", strength or "", 1, 10))
    if normalized_strength:
        # A bare "500" ("napa 500 tablet") is a strength the cashier left the unit
        # off of. It matches only the digits written on the row: 500 does not stand
        # in for 650, and a row with no strength at all keeps the number for its
        # name to answer.
        strength_numbers = frozenset(_STRENGTH_NUMBER_RE.findall(normalized_strength))
        for number in _bare_number_terms(remaining):
            if number not in strength_numbers:
                return None
            remaining = _remove_phrase(remaining, number)
            support.append(MedicineMatch("strength", "supporting", strength or "", 1, 10))
    if not remaining:
        return support[0] if support else None
    if normalized_name == remaining:
        return MedicineMatch("name", "exact", name, 1, 2)
    if normalized_generic and normalized_generic == remaining:
        return MedicineMatch("genericName", "exact", generic_name or "", 1, 3)
    if remaining in normalized_name:
        return MedicineMatch("name", "partial", name, len(remaining) / len(normalized_name), 5)
    if normalized_generic and remaining in normalized_generic:
        return MedicineMatch("genericName", "partial", generic_name or "", len(remaining) / len(normalized_generic), 6)
    name_score = _fuzzy_score(remaining, normalized_name)
    generic_score = _fuzzy_score(remaining, normalized_generic) if normalized_generic else None
    if name_score is not None:
        return MedicineMatch("name", "fuzzy", name, name_score, 8)
    if generic_score is not None:
        return MedicineMatch("genericName", "fuzzy", generic_name or "", generic_score, 9)
    return None


def best_match(*matches: MedicineMatch | None) -> MedicineMatch | None:
    available = [match for match in matches if match is not None]
    return min(available, key=lambda match: (match.rank, -match.score)) if available else None
